source/extensions/upstreams/http/udp/upstream_request.cc

```diff
@@ -0,0 +1,141 @@
+  Network::SocketPtr socket = createSocket(host_);
+  const Network::ConnectionInfoProvider& connection_info_provider =
+      socket->connectionInfoProvider();
+  ASSERT(upstream_to_downstream.connection().has_value());
+  Event::Dispatcher& dispatcher = upstream_to_downstream.connection()->dispatcher();
+  auto upstream =
+      std::make_unique<UdpUpstream>(&upstream_to_downstream, std::move(socket), host_, dispatcher);
+  StreamInfo::StreamInfoImpl stream_info(dispatcher.timeSource(), nullptr);
+  callbacks->onPoolReady(std::move(upstream), host_, connection_info_provider, stream_info, {});
+}
+
+UdpUpstream::UdpUpstream(Router::UpstreamToDownstream* upstream_to_downstream,
+                         Network::SocketPtr socket, Upstream::HostConstSharedPtr host,
+                         Event::Dispatcher& dispatcher)
+    : upstream_to_downstream_(upstream_to_downstream), socket_(std::move(socket)), host_(host),
+      dispatcher_(dispatcher) {
+  socket_->ioHandle().initializeFileEvent(
+      dispatcher_, [this](uint32_t) { onSocketReadReady(); }, Event::PlatformDefaultTriggerType,
+      Event::FileReadyType::Read);
+}
+
+void UdpUpstream::encodeData(Buffer::Instance& data, bool /*end_stream*/) {
+  for (const Buffer::RawSlice& slice : data.getRawSlices()) {
+    absl::string_view mem_slice(static_cast<const char*>(slice.mem_), slice.len_);
+    if (!capsule_parser_.IngestCapsuleFragment(mem_slice)) {
+      ENVOY_LOG_MISC(error, "Capsule ingestion error occured: slice = {}", mem_slice);
+      break;
+    }
+  }
+  capsule_parser_.ErrorIfThereIsRemainingBufferedData();
+}
+
+Envoy::Http::Status UdpUpstream::encodeHeaders(const Envoy::Http::RequestHeaderMap&,
+                                               bool /*end_stream*/) {
+  Api::SysCallIntResult rc = socket_->connect(host_->address());
+  if (SOCKET_FAILURE(rc.return_value_)) {
+    return absl::InternalError("Upstream socket connect failure.");
+  }
+  // Synthesize the 200 response headers downstream to complete the CONNECT-UDP handshake.
+  Envoy::Http::ResponseHeaderMapPtr headers{
+      Envoy::Http::createHeaderMap<Envoy::Http::ResponseHeaderMapImpl>(
+          {{Envoy::Http::Headers::get().Status, "200"},
+           {Envoy::Http::Headers::get().CapsuleProtocol, "?1"}})};
+  upstream_to_downstream_->decodeHeaders(std::move(headers), false);
+  return Envoy::Http::okStatus();
+}
+
+void UdpUpstream::resetStream() { upstream_to_downstream_ = nullptr; }
+
+void UdpUpstream::onSocketReadReady() {
+  uint32_t packets_dropped = 0;
+  const Api::IoErrorPtr result = Network::Utility::readPacketsFromSocket(
+      socket_->ioHandle(), *socket_->connectionInfoProvider().localAddress(), *this,
+      dispatcher_.timeSource(), /*prefer_gro=*/false, packets_dropped);
+  if (result == nullptr) {
+    socket_->ioHandle().activateFileEvents(Event::FileReadyType::Read);
+    return;
+  }
+}
+
+void UdpUpstream::processPacket(Network::Address::InstanceConstSharedPtr /*local_address*/,
```

Comment:
To confirm: we're ignoring `local_address` and `peer_address` here because we're using a bound and connected socket?